agent/runner.go

```diff
@@ -123,76 +122,66 @@ func (r *Runner) Run(runnerCtx context.Context) error { //nolint:contextcheck
 			case <-time.After(time.Minute):
 				logger.Debug().Msg("pipeline lease renewed")
 
-				if err := r.client.Extend(workflowCtx, work.ID); err != nil {
+				if err := r.client.Extend(workflowCtx, workflow.ID); err != nil {
 					log.Error().Err(err).Msg("extending pipeline deadline failed")
 				}
 			}
 		}
 	}()
 
-	state := rpc.State{}
+	state := rpc.WorkflowState{}
 	state.Started = time.Now().Unix()
 
-	err = r.client.Init(runnerCtx, work.ID, state)
+	err = r.client.Init(runnerCtx, workflow.ID, state)
 	if err != nil {
-		logger.Error().Err(err).Msg("pipeline initialization failed")
+		logger.Error().Err(err).Msg("workflow initialization failed")
+		// TODO: should we return here?
 	}
 
 	var uploads sync.WaitGroup
 	//nolint:contextcheck
-	err = pipeline.New(work.Config,
+	err = pipeline.New(workflow.Config,
 		pipeline.WithContext(workflowCtx),
-		pipeline.WithTaskUUID(fmt.Sprint(work.ID)),
-		pipeline.WithLogger(r.createLogger(logger, &uploads, work)),
-		pipeline.WithTracer(r.createTracer(ctxMeta, logger, work)),
+		pipeline.WithTaskUUID(fmt.Sprint(workflow.ID)),
+		pipeline.WithLogger(r.createLogger(logger, &uploads, workflow)),
+		pipeline.WithTracer(r.createTracer(ctxMeta, logger, workflow)),
 		pipeline.WithBackend(*r.backend),
 		pipeline.WithDescription(map[string]string{
-			"ID":       work.ID,
+			"ID":       workflow.ID,
 			"Repo":     repoName,
 			"Pipeline": pipelineNumber,
 		}),
 	).Run(runnerCtx)
 
 	state.Finished = time.Now().Unix()
-	state.Exited = true
-
-	if canceled.IsSet() {
-		state.Error = ""
-		state.ExitCode = pipeline.ExitCodeKilled
-	} else if err != nil {
-		pExitError := &pipeline.ExitError{}
-		switch {
-		case errors.As(err, &pExitError):
-			state.ExitCode = pExitError.Code
-		case errors.Is(err, pipeline.ErrCancel):
-			state.Error = ""
-			state.ExitCode = pipeline.ExitCodeKilled
-			canceled.SetTo(true)
-		default:
-			state.ExitCode = 1
-			state.Error = err.Error()
-		}
+
+	if errors.Is(err, pipeline.ErrCancel) {
+		canceled = true
+	} else if canceled {
+		err = errors.Join(err, pipeline.ErrCancel)
+	}
+
+	if err != nil {
+		state.Error = err.Error()
```

Comment:
Note for reviewers: this is the only bigger change in this PR. As workflow state updates never considered exit-codes in the actual logic and just used them for logging, I've removed them. Instead `state.Error` is now set respectively.